File: experiments/logistic_regression.py
"""
@file logistic_regression.py
@author Ryan Curtin

Use scipy to optimize the logistic regression objective function using L-BFGS
and automatic differentation.
"""
import numpy as np
import scipy.optimize
import random
import sys

if len(sys.argv) != 5:
  print("args: <trainFile> <trainLabelsFile> <testFile> <testLabelsFile>")
  exit(1)

trainFile = sys.argv[1]
trainLabelsFile = sys.argv[2]
testFile = sys.argv[3]
testLabelsFile = sys.argv[4]

trainData = np.genfromtxt(trainFile, delimiter=',', dtype=np.float64)
trainLabels = np.genfromtxt(trainLabelsFile, delimiter=',', dtype=np.float64)
testData = np.genfromtxt(testFile, delimiter=',', dtype=np.float64)
testLabels = np.genfromtxt(testLabelsFile, delimiter=',', dtype=np.float64)

def f(theta):
  objReg = 0.5 / 2.0 * np.dot(theta[1:], theta[1:])
  sigmoids = 1.0 / (1.0 + np.exp(np.minimum(300.0, -(theta[0] + np.matmul(trainData, theta[1:])))))
  innerSecondTerm = 1.0 - trainLabels + np.multiply(sigmoids, (2.0 * trainLabels - 1.0))
  result = np.sum(np.log(innerSecondTerm + 1e-10))
  return objReg - result

def g(theta):
  sigmoids = 1.0 / (1.0 + np.exp(-(theta[0] + np.matmul(trainData, theta[1:]))))
  G = np.zeros(len(theta))
  G[0] = -np.sum(trainLabels - sigmoids)
  G[1:] = (np.matmul((sigmoids - trainLabels), trainData) + 0.5 * theta[1:])
  return G

def compute_accuracy(theta, X, y):
  predictions = np.floor(1.0 / (1.0 + np.exp(-theta[0] - np.matmul(X,
      theta[1:]))) + 0.5)
  return np.sum(np.equal(predictions, y)) / len(y)

theta = np.zeros(np.shape(trainData)[1] + 1)

import timeit

def run():
  # Try to match mlpack configuration.
  return scipy.optimize.fmin_l_bfgs_b(f, theta, fprime=g, maxiter=10, maxls=50, epsilon=0.0, factr=0.0)

t = timeit.timeit(stmt=run, number=1)
print(t)

finalTheta, obj, d = run()
print(d)

print("Training accuracy:")
print(compute_accuracy(finalTheta, trainData, trainLabels))
print("Test accuracy:")
print(compute_accuracy(finalTheta, testData, testLabels))

File: experiments/logistic_regression.m
% This is actually a GNU Octave file because I don't have access to the Global
% Optimization Toolkit.
pkg load optim;

global trainData;
global trainLabels;
global testData;
global testLabels;

function [obj, grad] = lr(theta)
  global trainData;
  global trainLabels;
  objReg = 0.5 / 2.0 * (theta(2:end)' * theta(2:end));
  sigmoids = 1.0 ./ (1.0 .+ exp(-(theta(1) .+ trainData * theta(2:end))));
  innerSecondTerm = 1.0 .- trainLabels .+ sigmoids .* (2.0 .* trainLabels .- 1.0);

  obj = objReg - sum(log(innerSecondTerm));
  grad = [-sum(trainLabels .- sigmoids); (((sigmoids - trainLabels)' * trainData)' .+ 0.5 .* theta(2:end))];
endfunction

function [acc] = compute_accuracy(theta, data, labels)
  predictions = floor(1.0 ./ (1.0 .+ exp(-(theta(1) .+ data * theta(2:end)))) .+ 0.5);

  acc = sum(predictions == labels) / size(labels)(1);
end

args = argv();
trainFile = args{1, 1};
trainLabelsFile = args{2, 1};
testFile = args{3, 1};
testLabelsFile = args{4, 1};

trainData = csvread(trainFile);
trainLabels = csvread(trainLabelsFile);
testData = csvread(testFile);
testLabels = csvread(testLabelsFile);

% I guess maybe there is a JIT or something that might accelerate this, so let's
% let it "burn in" once.
theta0 = zeros(size(trainData)(2) + 1, 1);
control = {10, 0}; % maxiters, verbosity
bfgsmin('lr', {theta0}, control);

tic;
[thetaOut, objVal, conv, iters] = bfgsmin('lr', {theta0}, control);
toc

compute_accuracy(thetaOut, trainData, trainLabels)
compute_accuracy(thetaOut, testData, testLabels)
